File: 1530-number-of-good-leaf-nodes-pairs/1530-number-of-good-leaf-nodes-pairs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = 0;
    vector<int> dfs(TreeNode* root, int dist) {
        if(!root) return {};
        if(!root->left && !root->right) return {1};
        vector<int>leftLeaf = dfs(root->left, dist);
        vector<int>rightLeaf = dfs(root->right, dist);
        for(int l:leftLeaf) {
            for(int r:rightLeaf) {
                if(l+r <= dist) ans++;
            }
        }
        vector<int>overAllLeaf;
        for(int l:leftLeaf) overAllLeaf.push_back(l+1);
        for(int r:rightLeaf) overAllLeaf.push_back(r+1);
        return overAllLeaf;
    }
    int countPairs(TreeNode* root, int distance) {
        dfs(root, distance);
        return ans;
    }
};

File: 1530-number-of-good-leaf-nodes-pairs/1530-number-of-good-leaf-nodes-pairs.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans = 0;
    List<Integer> dfs(TreeNode root, int dist) {
        if(root == null) return new ArrayList<>();
        if(root.left == null && root.right == null) return new ArrayList<>(Arrays.asList(1));
        List<Integer>leftLeaf = dfs(root.left, dist);
        List<Integer>rightLeaf = dfs(root.right, dist);
        for(int x:leftLeaf) {
            for(int y:rightLeaf) {
                if(x+y <= dist) ans++;
            }
        }
        List<Integer>overAllLeaf = new ArrayList<>();
        for(int x:leftLeaf) overAllLeaf.add(x + 1);
        for(int x:rightLeaf) overAllLeaf.add(x+1);
        return overAllLeaf;
    }
    public int countPairs(TreeNode root, int distance) {
        dfs(root, distance);
        return ans;
    }
}